File: day13/.ts
type Factory = string[];
type Result = 'completed' | 'broken' | 'loop';

function runFactory(factory: Factory): Result {
  let pos = { row: 0, col: 0 };
  let walkedOn = new Set<string>();
  
  while (true) {
    if (pos.row < 0 || pos.row >= factory.length || pos.col < 0 || pos.col >= factory[0].length) {
      return "broken";
    }
    const posKey = pos.row.toString() + pos.col.toString()
    
    if (walkedOn.has(posKey)) return "loop";

    const move = factory[pos.row][pos.col];
    if (move === ".") return "completed";

    walkedOn.add(posKey);

    if (move === "<") pos.col--;
    else if (move === ">") pos.col++;
    else if (move === "^") pos.row--;
    else if (move === "v") pos.row++;
    else return "broken";
  }
}

File: day13/.py
def run_factory(factory: list[str]) -> str:
  pos = { "row": 0, "col": 0 };
  walked_on = set()
  
  while True:
    if pos['row'] < 0 or pos['row'] >= len(factory) or pos['col'] < 0 or pos['col'] >= len(factory[0]):
      return "broken"

    pos_key = f"{pos['row']},{pos['col']}"
    
    if pos_key in walked_on:
      return "loop"

    move = factory[pos['row']][pos['col']]
    if move == ".":
      return "completed"

    walked_on.add(pos_key)

    if move == "<":
      pos['col'] -= 1
    elif move == ">":
      pos['col'] += 1
    elif move == "^":
      pos['row'] -= 1
    elif move == "v":
      pos['row'] += 1
    else:
      return "broken"
